File: project/indexer/compiledb.py
"""indexer.compiledb -- load compile_commands.json and strip args for parse().

Wraps clang's CompilationDatabase (lab section 4.5): the raw command contains
the driver token, the source filename, and -c/-o pairs that libclang.parse()
must NOT see; relative -I paths are resolved against the command's directory.
"""

from __future__ import annotations

import os
import re
import tempfile

from clang.cindex import CompilationDatabase

from indexer import pathx as _pathx
from indexer.pathx import resolve_fs_path  # noqa: F401
from indexer.pathx import split_base_version  # re-exported for CLI use  # noqa: F401


def commands_from_text(text: str):
    """Parse compile_commands.json text into clang compile-command objects.

    Accepts either a single entry object ('{...}') or a full array ('[...]');
    a lone object is wrapped in an array. The text is written to a throwaway
    compile_commands.json and loaded through the same CompilationDatabase path
    `import` uses, so `cidx file -import-args` strips args identically. Each
    entry needs `directory`, `file`, and `arguments` (or `command`)."""
    body = text.lstrip()
    if body.startswith("{"):
        text = "[" + text + "]"
    with tempfile.TemporaryDirectory() as d:
        with open(os.path.join(d, "compile_commands.json"), "w") as fh:
            fh.write(text)
        return load_commands(d)


def db_directory(db_path: str) -> str:
    """Absolute directory holding compile_commands.json (db_path is the file or its dir)."""
    db_dir = (
        db_path[: -len("compile_commands.json")] or "."
        if db_path.endswith("compile_commands.json")
        else db_path
    )
    return os.path.abspath(db_dir)


def load_commands(db_path: str):
    """All compile commands from a compile_commands.json (file or its directory)."""
    cdb = CompilationDatabase.fromDirectory(db_directory(db_path))
    return list(cdb.getAllCompileCommands())


def _abs(p: str, base: str) -> str:
    return p if os.path.isabs(p) else os.path.normpath(os.path.join(base, p))


#: A leading token of the form NAME=value is an environment-variable assignment
#: (e.g. CCACHE_DIR=/x, CCACHE_AOPS-52378DIR=). The name must start with a
#: letter/underscore so a flag like -DFOO=bar (starts with '-') is never matched.
_ENV_ASSIGN_RE = re.compile(r"^[A-Za-z_][A-Za-z0-9_-]*=", re.ASCII)

#: Compiler-launcher wrappers that sit BEFORE the real compiler in a command
#: (matched by basename). libclang must see neither these nor the env
#: assignments — clang reports them as "'linker' input unused", and the driver
#: must point at the actual compiler.
_LAUNCHERS = frozenset(
    {"ccache", "sccache", "distcc", "icecc", "icerun", "env", "time", "nice"}
)


def command_start(args: list[str]) -> int:
    """Index of the real compiler driver in a raw command vector.

    A compile_commands.json `command` may be prefixed with environment-variable
    assignments (CCACHE_DIR=..., FOO=bar) and compiler-launcher wrappers
    (ccache, sccache, distcc, icecc, env, time, nice) before the actual
    compiler. Returns the index of the first token that is neither an env
    assignment nor a known launcher (0 for a plain `cc ...`). If the whole
    vector is prefix (degenerate), returns 0 so callers stay in bounds.
    """
    i = 0
    n = len(args)
    while i < n:
        tok = args[i]
        if _ENV_ASSIGN_RE.match(tok):
            i += 1
            continue
        if os.path.basename(tok) in _LAUNCHERS:
            i += 1
            continue
        break
    return i if i < n else 0


#: Flags libclang must not see, beyond the driver/source/-c/-o basics.
#: Dependency generation (-M*) writes build artifacts -- the obj/dep dir
#: usually does not exist outside a real build, which surfaces as a fatal
#: "error opening '...'" diagnostic. -Werror (and friends) promote benign
#: warnings to errors, and the indexer treats error diagnostics as a failed
#: parse -- a warning gcc never emitted must not abort indexing under clang.
#:
#: We also drop everything that only affects LINKING, CODEGEN-output, or the
#: MODULE/diagnostic CACHE: a compile_commands.json entry carries the full
#: driver invocation (link libraries, linker scripts, the module cache dir,
#: assembler passthrough ...), none of which influence the AST a libclang
#: *parse* produces -- and several of them (a missing -L dir, a -Wl, group,
#: a stale module cache) only add noise or fatal diagnostics. Header-search
#: flags (-nostdinc / -nostdinc++) and preprocessor-affecting flags
#: (-pthread, -fPIC) are NOT linker flags and are deliberately KEPT.
_DROP = frozenset(
    {
        "-c",
        "--",
        # -- dependency generation --
        "-M",
        "-MM",
        "-MD",
        "-MMD",
        "-MG",
        "-MP",
        "-MV",
        # -- warnings-as-errors --
        "-Werror",
        "-pedantic-errors",
        # -- link-stage modes (no effect on parsing) --
        "-shared",
        "-static",
        "-rdynamic",
        "-pie",
        "-no-pie",
        "-s",  # strip symbols at link
        "-pipe",  # use pipes instead of temp files (codegen plumbing)
        "-nostdlib",
        "-nodefaultlibs",
        "-nostartfiles",
        "-static-libgcc",
        "-shared-libgcc",
        "-static-libstdc++",
    }
)
_DROP_WITH_ARG = frozenset(
    {
        "-o",
        "-MF",
        "-MT",
        "-MQ",
        "-dependency-file",
        "--serialize-diagnostics",
        # -- linker options that take a separate argument --
        "-Xlinker",
        "-T",  # linker script
        "-L",  # space form: -L /usr/lib
        "-l",  # space form: -l foo
    }
)
_DROP_PREFIX = (
    "-Werror=",  # -Werror=return-type: keep it a plain warning
    "-Wp,-M",  # -Wp,-MD,<file> / -Wp,-MMD,<file>
    "-MF",
    "-MT",
    "-MQ",  # glued forms: -MF<file> etc.
    # -- linker / library / cache (glued forms) --
    "-l",  # -lfoo  (no frontend flag starts with -l)
    "-L",  # -L/usr/lib (no frontend flag starts with -L)
    "-Wl,",  # linker passthrough
    "-Wa,",  # assembler passthrough
    "-fuse-ld=",  # linker selection
    "-fmodules-cache-path=",  # module/diagnostic cache dir
)


def sanitize(args: list[str]) -> list[str]:
    """Re-apply the drop rules to already-stored options.

    Options are stripped at import time, so an index written by an older
    version may still carry flags the current rules would drop (-Werror,
    -MF ...); sanitizing again at parse time heals such databases without
    a re-import.

    Also heals a command PREFIX an older import stored when argv[0] was an
    env-var assignment rather than the compiler (e.g. CCACHE_DIR=... was
    dropped, leaving ["CCACHE_COMPRESS=1", ..., "ccache", "g++", "-g", ...]).
    When the first stored token is an env assignment or a launcher, everything
    up to and including the real compiler is dropped. (A re-import is still the
    proper fix — it also corrects the stored driver field.)
    """
    items = list(args)
    if items and (
        _ENV_ASSIGN_RE.match(items[0])
        or os.path.basename(items[0]) in _LAUNCHERS
    ):
        items = items[command_start(items) + 1 :]
    out: list[str] = []
    it = iter(items)
    for tok in it:
        if tok in _DROP:
            continue
        if tok in _DROP_WITH_ARG:
            next(it, None)
            continue
        if tok.startswith(_DROP_PREFIX):
            continue
        out.append(tok)
    return out


def _is_indirected(val: str) -> bool:
    """True if a -I/-isystem/-iquote value already contains a label or env-var.

    Such values must be preserved verbatim (not absolutized) so the stored
    indirected form survives import unchanged and resolves at parse time.
    """
    return "<" in val or "$" in val


def strip_for_libclang(cmd) -> list[str]:
    """Raw driver invocation -> flags parse() wants. Resolves relative includes.

    Preserve rule (v14): if a -I/-isystem/-iquote value already contains '<'
    or '$', emit it verbatim (do NOT absolutize). Otherwise keep the existing
    absolutize-relative behaviour.
    """
    raw, directory = list(cmd.arguments), cmd.directory
    src = {cmd.filename, os.path.basename(cmd.filename)}
    out: list[str] = []
    # Drop the whole command prefix: leading env-var assignments + launcher
    # wrappers (ccache ...) AND the real compiler token at command_start.
    it = iter(raw[command_start(raw) + 1 :])
    for tok in it:
        if tok in _DROP:
            continue
        if tok in _DROP_WITH_ARG:
            next(it, None)  # drop flag + its argument
            continue
        if tok.startswith(_DROP_PREFIX):
            continue
        if tok in src:
            continue
        matched = False
        for flag in ("-I", "-isystem", "-iquote"):
            if tok == flag:  # space form: -I path
                val = next(it, "")
                if _is_indirected(val):
                    out += [flag, val]  # preserve verbatim
                else:
                    out += [flag, _abs(val, directory)]
                matched = True
                break
            if tok.startswith(flag) and len(tok) > len(flag):  # glued: -Ipath
                val = tok[len(flag) :]
                if _is_indirected(val):
                    out.append(tok)  # preserve verbatim
                else:
                    out.append(flag + _abs(val, directory))
                matched = True
                break
        if not matched:
            out.append(tok)
    return out


def source_path(cmd) -> str:
    """Absolute path of the command's source file."""
    return _abs(cmd.filename, cmd.directory)


def driver(cmd) -> str:
    """The command's compiler driver (argv[0]).

    A custom-toolchain driver (e.g. /opt/1A/toolchain/.../bin/g++) carries its
    own header search paths; storing it lets parse() replicate them later.
    Bare names ('cc', 'g++') are kept as-is and resolved via PATH at parse
    time; relative paths are resolved against the command's directory. Leading
    env-var assignments and launcher wrappers (CCACHE_DIR=... ccache g++) are
    skipped so the REAL compiler is returned, not the env prefix.
    """
    args = list(cmd.arguments)
    if not args:
        return ""
    argv0 = args[command_start(args)]
    return _abs(argv0, cmd.directory) if os.sep in argv0 else argv0


# ---------------------------------------------------------------------------
# Include-path aliasing (v0.6.0): encode absolute -I dirs <-> <label> tokens.
#
# Two paired transforms over the -I/-isystem/-iquote VALUES only (both the
# 'space' form `-I path` and the 'glued' form `-Ipath`); every other token
# passes through untouched:
#
#   alias_options   ENCODE  absolute dir  -> <label>   (registry, longest match)
#   resolve_options DECODE  <label>/$VAR  -> absolute  (parse-time, for libclang)
#
# Stored compile_options keep the indirected (aliased) form so the index is
# portable; resolve_options is applied just before handing args to libclang.
# ---------------------------------------------------------------------------

_INCLUDE_FLAGS = ("-I", "-isystem", "-iquote")


def _map_include_values(options, fn) -> list[str]:
    """Return a copy of *options* with *fn* applied to each include-path value.

    Handles `-I path` (two tokens) and `-Ipath` (glued); all other tokens are
    copied verbatim. fn receives the bare path value and returns its rewrite.
    """
    out: list[str] = []
    it = iter(options)
    for tok in it:
        matched = False
        for flag in _INCLUDE_FLAGS:
            if tok == flag:  # space form: -I path
                out += [flag, fn(next(it, ""))]
                matched = True
                break
            if tok.startswith(flag) and len(tok) > len(flag):  # glued: -Ipath
                out.append(flag + fn(tok[len(flag) :]))
                matched = True
                break
        if not matched:
            out.append(tok)
    return out


def include_values(options):
    """Yield each include-path VALUE from *options* (both `-I path` and `-Ipath`
    forms), skipping every other token. Read-only counterpart of
    _map_include_values, used by the import version-bump scan."""
    it = iter(options)
    for tok in it:
        for flag in _INCLUDE_FLAGS:
            if tok == flag:
                yield next(it, "")
                break
            if tok.startswith(flag) and len(tok) > len(flag):
                yield tok[len(flag) :]
                break


def resolve_options(options, lookup=None, autoderive: bool = True) -> list[str]:
    """DECODE: resolve <label>/$VAR/~ in include-path values to absolute paths.

    Only values that look indirected (contain '<' or '$' or start with '~') are
    resolved (via the full resolution chain + abspath); already-absolute plain
    paths are left untouched. Used at parse/index time so libclang sees real
    directories.
    """

    def _fn(val: str) -> str:
        if "<" in val or "$" in val or val.startswith("~"):
            return os.path.abspath(
                resolve_fs_path(val, lookup=lookup, autoderive=autoderive)
            )
        return val

    return _map_include_values(options, _fn)


def build_label_map(labels, lookup=None) -> list[tuple[str, str, bool]]:
    """Build the encode lookup from (name, stored_path[, versioned]) entries.

    Accepts 2-tuples (label, exact match) or 3-tuples (the *versioned* flag, set
    for components — version-agnostic match). Each stored path is resolved to an
    absolute directory (env-vars expanded, NO autoderive). Returned as
    (name, resolved_path, versioned) sorted most-specific first (longest
    resolved path, then name) so the longest prefix wins deterministically in
    both Python and C++.
    """
    out: list[tuple[str, str, bool]] = []
    for entry in labels:
        if len(entry) == 3:
            name, stored, versioned = entry
        else:
            name, stored = entry
            versioned = False
        rp = os.path.abspath(resolve_fs_path(stored, lookup=lookup, autoderive=False))
        out.append((name, rp, versioned))
    out.sort(key=lambda nr: (-len(nr[1]), nr[0]))
    return out


def match_alias(absval, label_map):
    """Longest-match an absolute path against *label_map*.

    Returns (name, version_segment, remainder) for the first (longest) entry
    whose resolved dir equals or contains *absval*, else None. For a *versioned*
    (component) entry the segment immediately after the base, if it looks like a
    version, is captured as *version_segment* and excluded from *remainder* (so
    the stored portable token carries NO version — it is re-injected at decode
    by get_alias -> base + highest version). For label entries version_segment
    is always None.
    """
    for name, rp, versioned in label_map:
        if absval == rp or absval.startswith(rp + os.sep):
            rem = absval[len(rp) :]
            vseg = None
            if versioned and rem.startswith(os.sep):
                head = rem[1:].split(os.sep, 1)
                if head[0] and _pathx.is_version_segment(head[0]):
                    vseg = head[0]
                    rem = (os.sep + head[1]) if len(head) > 1 else ""
            return name, vseg, rem
    return None


def alias_options(options, label_map) -> list[str]:
    """ENCODE: rewrite absolute include-path values to <label> tokens.

    *label_map* is the output of build_label_map (sorted longest-first). A value
    equal to or under an entry's resolved directory becomes `<name>` + remainder
    (longest match wins; for component entries the version segment is stripped).
    Values already indirected ('<' or '$') and relative values are unchanged.
    """

    def _fn(val: str) -> str:
        if "<" in val or "$" in val or not os.path.isabs(val):
            return val
        m = match_alias(os.path.normpath(val), label_map)
        if m is None:
            return val
        name, _vseg, rem = m
        return "<" + name + ">" + rem

    return _map_include_values(options, _fn)

File: cidx-cpp/src/compiledb/compiledb.cpp
#include "compiledb/compiledb.hpp"

#include <algorithm>
#include <cstring>
#include <functional>
#include <optional>
#include <regex>
#include <set>
#include <string>
#include <utility>

#include "clangx/libclang.hpp"
#include "util/errors.hpp"
#include "util/pathutil.hpp"

namespace cidx {
namespace {

// Frozen drop sets (compiledb.py, G10). Rationale baked into the Python
// comments: -M* writes build artifacts into dirs that don't exist outside a
// real build (fatal "error opening '...'"); -Werror promotes warnings gcc
// never emitted into clang error diagnostics. We also drop everything that
// only affects LINKING, CODEGEN-output, or the MODULE/diagnostic CACHE — none
// of it influences the AST a libclang *parse* produces. Header-search flags
// (-nostdinc / -nostdinc++) and preprocessor-affecting flags (-pthread, -fPIC)
// are NOT linker flags and are deliberately KEPT.
const std::set<std::string> kDrop = {
    "-c",       "--",
    // dependency generation
    "-M",       "-MM",      "-MD",      "-MMD", "-MG", "-MP", "-MV",
    // warnings-as-errors
    "-Werror",  "-pedantic-errors",
    // link-stage modes (no effect on parsing)
    "-shared",  "-static",  "-rdynamic", "-pie", "-no-pie", "-s", "-pipe",
    "-nostdlib", "-nodefaultlibs", "-nostartfiles",
    "-static-libgcc", "-shared-libgcc", "-static-libstdc++",
};
const std::set<std::string> kDropWithArg = {
    "-o", "-MF", "-MT", "-MQ", "-dependency-file", "--serialize-diagnostics",
    // linker options that take a separate argument
    "-Xlinker", "-T", "-L", "-l",
};
const char *const kDropPrefix[] = {
    "-Werror=", // -Werror=return-type: keep it a plain warning
    "-Wp,-M",   // -Wp,-MD,<file> / -Wp,-MMD,<file>
    "-MF",      // glued forms: -MF<file> etc.
    "-MT",
    "-MQ",
    // linker / library / cache (glued forms)
    "-l",     // -lfoo  (no frontend flag starts with -l)
    "-L",     // -L/usr/lib (no frontend flag starts with -L)
    "-Wl,",   // linker passthrough
    "-Wa,",   // assembler passthrough
    "-fuse-ld=",            // linker selection
    "-fmodules-cache-path=", // module/diagnostic cache dir
};

bool has_drop_prefix(const std::string &tok) {
  for (const char *prefix : kDropPrefix) {
    if (tok.starts_with(prefix)) {
      return true;
    }
  }
  return false;
}

// A leading NAME=value token is an env-var assignment; the name must start with
// a letter/underscore so a flag like -DFOO=bar (starts '-') is never matched.
bool is_env_assignment(const std::string &tok) {
  static const std::regex kEnvAssign(R"(^[A-Za-z_][A-Za-z0-9_-]*=)");
  return std::regex_search(tok, kEnvAssign);
}

// Compiler-launcher wrappers that sit before the real compiler (by basename).
const std::set<std::string> kLaunchers = {
    "ccache", "sccache", "distcc", "icecc", "icerun", "env", "time", "nice",
};

bool is_launcher(const std::string &tok) {
  return kLaunchers.count(pathutil::basename(tok)) != 0;
}

// compiledb.py:23-24 — absolute paths returned UNCHANGED (not normalized).
std::string abs_against(const std::string &p, const std::string &base) {
  if (pathutil::isabs(p)) {
    return p;
  }
  return pathutil::normpath(pathutil::join(base, p));
}

std::string to_string(const LibClang &lib, CXString s) {
  return CxString(lib, s).str();
}

} // namespace

std::string CompileDb::db_dir_from_arg(const std::string &db_arg) {
  static const std::string kJson = "compile_commands.json";
  if (db_arg.size() >= kJson.size() &&
      db_arg.compare(db_arg.size() - kJson.size(), kJson.size(), kJson) == 0) {
    std::string dir = db_arg.substr(0, db_arg.size() - kJson.size());
    return dir.empty() ? "." : dir; // `db_path[:-len(...)] or "."`
  }
  return db_arg;
}

std::vector<CompileCommand> CompileDb::load(const std::string &db_arg) {
  LibClang &lib = LibClang::instance();
  lib.load();

  const std::string abs_dir = pathutil::abspath(db_dir_from_arg(db_arg));
  CXCompilationDatabase_Error error = CXCompilationDatabase_NoError;
  CXCompilationDatabase db =
      lib.clang_CompilationDatabase_fromDirectory(abs_dir.c_str(), &error);
  if (error != CXCompilationDatabase_NoError || db == nullptr) {
    if (db != nullptr) {
      lib.clang_CompilationDatabase_dispose(db);
    }
    throw CidxError("could not load compilation database from '" + abs_dir +
                    "'");
  }

  std::vector<CompileCommand> out;
  CXCompileCommands cmds =
      lib.clang_CompilationDatabase_getAllCompileCommands(db);
  if (cmds != nullptr) {
    const unsigned n = lib.clang_CompileCommands_getSize(cmds);
    for (unsigned i = 0; i < n; ++i) {
      CXCompileCommand cmd = lib.clang_CompileCommands_getCommand(cmds, i);
      CompileCommand cc;
      cc.directory =
          to_string(lib, lib.clang_CompileCommand_getDirectory(cmd));
      cc.filename = to_string(lib, lib.clang_CompileCommand_getFilename(cmd));
      std::vector<std::string> raw;
      const unsigned argc = lib.clang_CompileCommand_getNumArgs(cmd);
      raw.reserve(argc);
      for (unsigned j = 0; j < argc; ++j) {
        raw.push_back(to_string(lib, lib.clang_CompileCommand_getArg(cmd, j)));
      }
      cc.driver = driver(raw, cc.directory);
      cc.args = strip_for_libclang(raw, cc.filename, cc.directory);
      out.push_back(std::move(cc));
    }
    lib.clang_CompileCommands_dispose(cmds);
  }
  lib.clang_CompilationDatabase_dispose(db);
  return out;
}

std::vector<std::string>
CompileDb::strip_for_libclang(const std::vector<std::string> &argv,
                              const std::string &filename,
                              const std::string &directory) {
  const std::set<std::string> src = {filename, pathutil::basename(filename)};
  std::vector<std::string> out;
  // Drop the whole command prefix: env-var assignments + launcher wrappers
  // (ccache ...) AND the real compiler token at command_start.
  size_t i = argv.empty() ? 0 : command_start(argv) + 1;
  while (i < argv.size()) {
    const std::string &tok = argv[i++];
    if (kDrop.count(tok) != 0) {
      continue;
    }
    if (kDropWithArg.count(tok) != 0) {
      if (i < argv.size()) {
        ++i; // drop flag + its argument
      }
      continue;
    }
    if (has_drop_prefix(tok)) {
      continue;
    }
    if (src.count(tok) != 0) {
      continue;
    }
    bool matched = false;
    for (const char *flag : {"-I", "-isystem", "-iquote"}) {
      const size_t flen = std::strlen(flag);
      if (tok == flag) { // space form: -I path
        const std::string arg = i < argv.size() ? argv[i++] : std::string();
        out.push_back(flag);
        // Preserve rule (portable-paths §5): if the value contains '<' or '$'
        // it is a template/env-var reference — emit verbatim, not absolutized.
        if (arg.find('<') != std::string::npos ||
            arg.find('$') != std::string::npos) {
          out.push_back(arg);
        } else {
          out.push_back(abs_against(arg, directory));
        }
        matched = true;
        break;
      }
      if (tok.size() > flen && tok.starts_with(flag)) { // glued
        const std::string val = tok.substr(flen);
        // Preserve rule: if the value portion contains '<' or '$', emit the
        // entire original token verbatim (flag + value, already glued).
        if (val.find('<') != std::string::npos ||
            val.find('$') != std::string::npos) {
          out.push_back(tok); // verbatim: e.g. "-I<libfoo-include>/include"
        } else {
          out.push_back(flag + abs_against(val, directory));
        }
        matched = true;
        break;
      }
    }
    if (!matched) {
      out.push_back(tok);
    }
  }
  return out;
}

std::vector<std::string>
CompileDb::sanitize(const std::vector<std::string> &stored) {
  std::vector<std::string> out;
  // Heal a command prefix an older import stored when argv[0] was an env-var
  // assignment rather than the compiler (e.g. ["CCACHE_COMPRESS=1", ...,
  // "ccache", "g++", "-g", ...]). When the first stored token is an env
  // assignment or launcher, drop through the real compiler at command_start.
  size_t i = 0;
  if (!stored.empty() &&
      (is_env_assignment(stored[0]) || is_launcher(stored[0]))) {
    i = command_start(stored) + 1;
  }
  while (i < stored.size()) {
    const std::string &tok = stored[i++];
    if (kDrop.count(tok) != 0) {
      continue;
    }
    if (kDropWithArg.count(tok) != 0) {
      if (i < stored.size()) {
        ++i;
      }
      continue;
    }
    if (has_drop_prefix(tok)) {
      continue;
    }
    out.push_back(tok);
  }
  return out;
}

size_t CompileDb::command_start(const std::vector<std::string> &args) {
  size_t i = 0;
  const size_t n = args.size();
  while (i < n) {
    const std::string &tok = args[i];
    if (is_env_assignment(tok) || is_launcher(tok)) {
      ++i;
      continue;
    }
    break;
  }
  return i < n ? i : 0;
}

std::string CompileDb::driver(const std::vector<std::string> &argv,
                              const std::string &directory) {
  if (argv.empty()) {
    return std::string();
  }
  // Skip env-assignment + launcher prefix; the real compiler is the driver.
  const std::string &argv0 = argv[command_start(argv)];
  if (!argv0.contains('/')) {
    return argv0; // bare name: PATH resolution at parse time
  }
  return abs_against(argv0, directory);
}

// Version detection regex: ^v?[0-9]+([._-][0-9]+)*$
// Explicit [0-9] (not \d) to be locale-independent (contract §2).
std::pair<std::string, std::string>
CompileDb::split_base_version(const std::string &root) {
  // Step 1: normpath.
  const std::string normed = pathutil::normpath(root);
  // Step 2: split → (base, seg).
  const auto [base, seg] = pathutil::split(normed);
  // Step 3: match seg against version regex.
  if (seg.empty() || base.empty() || base == "/") {
    return {normed, ""};
  }
  // ^v?[0-9]+([._-][0-9]+)*$
  static const std::regex kVersionRe(R"(^v?[0-9]+([._\-][0-9]+)*$)");
  if (std::regex_match(seg, kVersionRe)) {
    return {base, seg};
  }
  return {normed, ""};
}

// ---------------------------------------------------------------------------
// Include-path aliasing (v0.6.0) — port of compiledb.py:
//   _map_include_values / resolve_options / build_label_map / alias_options
// ---------------------------------------------------------------------------

namespace {

// The three include flags (compiledb.py _INCLUDE_FLAGS).
const char *const kIncludeFlags[] = {"-I", "-isystem", "-iquote"};

// Apply fn to each include-path value in options (both space and glued forms).
// All other tokens are copied verbatim.
// Mirrors compiledb.py:_map_include_values.
template <typename Fn>
std::vector<std::string>
map_include_values(const std::vector<std::string> &options, Fn fn) {
  std::vector<std::string> out;
  out.reserve(options.size());
  size_t i = 0;
  while (i < options.size()) {
    const std::string &tok = options[i++];
    bool matched = false;
    for (const char *flag : kIncludeFlags) {
      const size_t flen = std::strlen(flag);
      if (tok == flag) { // space form: -I path
        out.push_back(tok);
        if (i < options.size()) {
          out.push_back(fn(options[i++]));
        }
        matched = true;
        break;
      }
      if (tok.size() > flen && tok.starts_with(flag)) { // glued: -Ipath
        out.push_back(std::string(flag) + fn(tok.substr(flen)));
        matched = true;
        break;
      }
    }
    if (!matched) {
      out.push_back(tok);
    }
  }
  return out;
}

} // namespace

// DECODE: resolve <label>/$VAR/~ in include-path values to absolute paths.
std::vector<std::string>
CompileDb::resolve_options(
    const std::vector<std::string> &options,
    std::function<std::optional<std::string>(const std::string &)> lookup,
    bool autoderive) {
  pathutil::LabelResolver resolver(lookup ? std::move(lookup) : nullptr,
                                   autoderive);
  return map_include_values(options, [&](const std::string &val) -> std::string {
    // Only resolve values that look indirected: contain '<', '$', or start
    // with '~'. Plain absolute paths pass through unchanged.
    if (val.find('<') == std::string::npos &&
        val.find('$') == std::string::npos &&
        (val.empty() || val[0] != '~')) {
      return val;
    }
    return pathutil::abspath(pathutil::resolve_fs_path(val, resolver));
  });
}

// Build the encode label map from (name, stored_path, versioned) entries.
std::vector<AliasEntry>
CompileDb::build_label_map(
    const std::vector<AliasEntry> &labels,
    std::function<std::optional<std::string>(const std::string &)> lookup) {
  // NO autoderive — a label's own stored value is taken literally.
  pathutil::LabelResolver resolver(lookup ? std::move(lookup) : nullptr,
                                   /*autoderive=*/false);
  std::vector<AliasEntry> out;
  out.reserve(labels.size());
  for (const auto &[name, stored, versioned] : labels) {
    const std::string rp =
        pathutil::abspath(pathutil::resolve_fs_path(stored, resolver));
    out.emplace_back(name, rp, versioned);
  }
  // Sort: longest resolved path first; on tie, by name ascending.
  std::sort(out.begin(), out.end(),
            [](const AliasEntry &a, const AliasEntry &b) {
              if (std::get<1>(a).size() != std::get<1>(b).size()) {
                return std::get<1>(a).size() > std::get<1>(b).size(); // longer
              }
              return std::get<0>(a) < std::get<0>(b); // then name ascending
            });
  return out;
}

// Longest-match an absolute path against label_map (compiledb.py:match_alias).
std::optional<std::tuple<std::string, std::string, std::string>>
CompileDb::match_alias(const std::string &absval,
                       const std::vector<AliasEntry> &label_map) {
  for (const auto &[name, rp, versioned] : label_map) {
    if (absval == rp || absval.starts_with(rp + "/")) {
      std::string rem = absval.substr(rp.size()); // "" or "/seg/..."
      std::string vseg;
      if (versioned && rem.starts_with("/")) {
        const std::string tail = rem.substr(1);
        const std::size_t slash = tail.find('/');
        const std::string head =
            slash == std::string::npos ? tail : tail.substr(0, slash);
        if (!head.empty() && is_version_segment(head)) {
          vseg = head;
          rem = slash == std::string::npos ? "" : tail.substr(slash);
        }
      }
      return std::make_tuple(name, vseg, rem);
    }
  }
  return std::nullopt;
}

// Yield each include-path value (compiledb.py:include_values).
std::vector<std::string>
CompileDb::include_values(const std::vector<std::string> &options) {
  std::vector<std::string> out;
  size_t i = 0;
  while (i < options.size()) {
    const std::string &tok = options[i++];
    for (const char *flag : kIncludeFlags) {
      const size_t flen = std::strlen(flag);
      if (tok == flag) {
        if (i < options.size()) {
          out.push_back(options[i++]);
        } else {
          out.emplace_back();
        }
        break;
      }
      if (tok.size() > flen && tok.starts_with(flag)) {
        out.push_back(tok.substr(flen));
        break;
      }
    }
  }
  return out;
}

// ENCODE: rewrite absolute include-path values to <label> tokens.
std::vector<std::string>
CompileDb::alias_options(const std::vector<std::string> &options,
                         const std::vector<AliasEntry> &label_map) {
  return map_include_values(options, [&](const std::string &val) -> std::string {
    // Already indirected, or relative: leave unchanged.
    if (val.find('<') != std::string::npos ||
        val.find('$') != std::string::npos ||
        !pathutil::isabs(val)) {
      return val;
    }
    const auto m = match_alias(pathutil::normpath(val), label_map);
    if (!m.has_value()) {
      return val;
    }
    return "<" + std::get<0>(*m) + ">" + std::get<2>(*m);
  });
}

// True iff seg matches the version regex (compiledb.py:is_version_segment).
bool CompileDb::is_version_segment(const std::string &seg) {
  static const std::regex kVersionRe(R"(^v?[0-9]+([._\-][0-9]+)*$)");
  return std::regex_match(seg, kVersionRe);
}

// Numeric per-segment version key (compiledb.py / pathx.version_key).
std::vector<long long> CompileDb::version_key(const std::string &version) {
  std::string v = version;
  if (!v.empty() && v[0] == 'v') {
    v = v.substr(1);
  }
  std::vector<long long> out;
  std::string cur;
  const auto flush = [&]() {
    if (!cur.empty()) {
      const bool digits = std::all_of(cur.begin(), cur.end(), [](char c) {
        return c >= '0' && c <= '9';
      });
      if (digits) {
        out.push_back(std::stoll(cur));
      }
      cur.clear();
    }
  };
  for (char c : v) {
    if (c == '.' || c == '_' || c == '-') {
      flush();
    } else {
      cur += c;
    }
  }
  flush();
  return out;
}

} // namespace cidx
